src/plugins/intel_npu/src/plugin/src/metadata.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "metadata.hpp"
+
+#include <cstring>
+#include <sstream>
+
+#include "intel_npu/config/config.hpp"
+#include "intel_npu/utils/logger/logger.hpp"
+#include "openvino/core/version.hpp"
+
+namespace intel_npu {
+
+OpenvinoVersion::OpenvinoVersion(const std::string& version) {
+    this->version = version;
+    this->size = static_cast<uint32_t>(version.size());
+}
+
+void OpenvinoVersion::read(std::istream& stream) {
+    stream.read(reinterpret_cast<char*>(&size), sizeof(size));
+    stream.read(&version[0], size);
+}
+
+Metadata<METADATA_VERSION_1_0>::Metadata() : ovVersion{ov::get_openvino_version().buildNumber} {
+    version = METADATA_VERSION_1_0;
+}
+
+void Metadata<METADATA_VERSION_1_0>::read(std::istream& stream) {
+    ovVersion.read(stream);
+}
+
+uint32_t& OpenvinoVersion::get_size() {
+    return size;
+}
+
+void Metadata<METADATA_VERSION_1_0>::write(std::ostream& stream) {
+    stream.write(reinterpret_cast<const char*>(&version), sizeof(version));
+    stream.write(reinterpret_cast<const char*>(&ovVersion.get_size()), sizeof(ovVersion.get_size()));
+    stream.write(ovVersion.get_version().data(), ovVersion.get_version().size());
+}
+
+std::unique_ptr<MetadataBase> create_metadata(uint32_t version) {
+    switch (version) {
+    case METADATA_VERSION_1_0:
+        return std::make_unique<Metadata<METADATA_VERSION_1_0>>();
+
+    default:
+        return nullptr;
+    }
+}
+
+std::string_view OpenvinoVersion::get_version() {
+    return std::string_view(version);
+}
+
+bool Metadata<METADATA_VERSION_1_0>::is_compatible() {
+    Logger _logger("NPUPlugin", Logger::global().level());
+    // checking if we can import the blob
+    if (ovVersion.get_version().compare(ov::get_openvino_version().buildNumber)) {
+        _logger.warning("Imported blob OpenVINO version: %s, but the current OpenVINO version is: %s",
+                        ovVersion.get_version().data(),
+                        ov::get_openvino_version().buildNumber);
+
+#ifdef NPU_PLUGIN_DEVELOPER_BUILD
+        if (auto envVar = std::getenv("NPU_DISABLE_VERSION_CHECK")) {
+            if (envVarStrToBool("NPU_DISABLE_VERSION_CHECK", envVar)) {
+                return true;
+            }
+        }
+#endif
+        return false;
+    }
+    return true;
+}
+
+std::unique_ptr<MetadataBase> read_metadata_from(std::vector<uint8_t>& blob) {
+    Logger _logger("NPUPlugin", Logger::global().level());
+    size_t magicBytesSize = MAGIC_BYTES.size();
+    std::string blobMagicBytes(magicBytesSize, '\0');
+
+    auto metadataIterator = blob.end() - magicBytesSize;
+    memcpy(blobMagicBytes.data(), &(*metadataIterator), magicBytesSize);
+    if (MAGIC_BYTES != blobMagicBytes) {
+        _logger.error("Blob is not versioned!");
```

Comment:
IMO it does not sound right to use throws in normal/expected cases, like this one.

Reply:
Doesn't sound valid to me, you are already throwing, but in the parent call instead of the current one. The difference is, you're placing the message in the logger and then throwing another message. My suggestions implied throwing and catcing in a hierarchical manner, which lead to concatenating this message and the one from the parent.

Also, we throw when we wish to halt execution, either because it is not possible or desired in the current state. This doesn't necessarily mean the state has to be unexpected, we may also throw in foreseen cases if execution cannot continue.

This one also holds little value, I've brought this suggestion since I believe that's how OV decided to handle exceptions. We don't fill the logger and then return a magic back to OV, instead we call their throw macros when we know we wanna stop.